bus-mapping/src/circuit_input_builder.rs

```diff
@@ -761,93 +787,172 @@ impl<'a> CircuitInputStateRef<'a> {
+                block
+                    .container
+                    .insert(Operation::new(block_ctx.rwc.inc_pre(), RW::WRITE, op));
+            }
+            OpEnum::TxAccessListAccount(op) => {
+                if !op.value {
+                    sdb.remove_account_from_access_list(&op.address);
+                }
+                block
+                    .container
+                    .insert(Operation::new(block_ctx.rwc.inc_pre(), RW::WRITE, op));
+            }
+            OpEnum::TxAccessListAccountStorage(op) => {
+                if !op.value {
+                    sdb.remove_account_storage_from_access_list(&(op.address, op.key));
+                }
+                block
+                    .container
+                    .insert(Operation::new(block_ctx.rwc.inc_pre(), RW::WRITE, op));
+            }
+            OpEnum::Account(op) => {
+                let (_, account) = sdb.get_account_mut(&op.address);
+                match op.field {
+                    AccountField::Nonce => account.nonce = op.value,
+                    AccountField::Balance => account.balance = op.value,
+                    AccountField::CodeHash => {
+                        account.code_hash = op.value.to_be_bytes().into();
                     }
                 }
+                block
+                    .container
+                    .insert(Operation::new(block_ctx.rwc.inc_pre(), RW::WRITE, op));
+            }
+            OpEnum::TxRefund(_) => unimplemented!(),
+            OpEnum::AccountDestructed(_) => unimplemented!(),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Handle a reversion caused by call_idx
+    fn handle_reversion(&mut self, call_idx: usize) {
+        let mut step_index = self.tx.steps.len() - 1;
+
+        loop {
+            let step = &self.tx.steps[step_index];
+
+            // Skip failure callees
+            if step.call_index != call_idx && !self.tx.calls[step.call_index].is_success {
+                step_index = self.tx_ctx.calls[step.call_index].step_index;
+                continue;
+            }
+
+            // Apply reversion
+            for op_ref in step.bus_mapping_instance.iter().rev() {
+                if let Some(op) = self.get_rev_op_by_ref(op_ref) {
+                    Self::apply_reverted_op(op, self.block, self.block_ctx, self.sdb);
+                }
+            }
+
+            // When we reach the step that triggers a call, assign the
+            // `rw_counter_end_of_reversion` to the call.
+            if CallKind::try_from(step.op).is_ok() {
+                let call_idx = self
+                    .tx_ctx
+                    .calls
+                    .iter()
+                    .position(|call| call.step_index == step_index)
+                    .expect("Call not found");
+                self.tx.calls[call_idx].rw_counter_end_of_reversion =
+                    self.block_ctx.rwc.0 - 1 - self.tx_ctx.calls[call_idx].swc_offset;
```

Comment:
I don't really understand the logic of this VS this:
```rust
let next_step = &self.tx.steps[step_index+1];
self.tx.calls[next_step.call_index].rw_counter_end_of_reversion = self.block_ctx.rwc.0 - 1;
```
The first part is about finding the call_index that corresponds to the call created in this step: since each step contains it's own call_index, the step following a call will contain the call_index of the call created by the step right?

The second part is about the `rw_counter_end_of_reversion` value.  At this point `self.block_ctx.rwc.0 - 1` points to the last reversed operation generated for the call created by the step.  Isn't that the value we should set for `rw_counter_end_of_reversion`?